File: app/src/main/java/com/limelight/nvstream/http/PairingManager.java
package com.limelight.nvstream.http;

import org.bouncycastle.crypto.BlockCipher;
import org.bouncycastle.crypto.engines.AESLightEngine;
import org.bouncycastle.crypto.params.KeyParameter;

import org.xmlpull.v1.XmlPullParserException;

import com.limelight.LimeLog;

import java.security.cert.Certificate;
import java.io.*;
import java.security.*;
import java.security.cert.*;
import java.util.Arrays;
import java.util.Locale;

public class PairingManager {

    private NvHTTP http;
    
    private PrivateKey pk;
    private X509Certificate cert;
    private byte[] pemCertBytes;

    private X509Certificate serverCert;
    
    public enum PairState {
        NOT_PAIRED,
        PAIRED,
        PIN_WRONG,
        FAILED,
        ALREADY_IN_PROGRESS
    }
    
    public PairingManager(NvHTTP http, LimelightCryptoProvider cryptoProvider) {
        this.http = http;
        this.cert = cryptoProvider.getClientCertificate();
        this.pemCertBytes = cryptoProvider.getPemEncodedClientCertificate();
        this.pk = cryptoProvider.getClientPrivateKey();
    }
    
    final private static char[] hexArray = "0123456789ABCDEF".toCharArray();
    private static String bytesToHex(byte[] bytes) {
        char[] hexChars = new char[bytes.length * 2];
        for ( int j = 0; j < bytes.length; j++ ) {
            int v = bytes[j] & 0xFF;
            hexChars[j * 2] = hexArray[v >>> 4];
            hexChars[j * 2 + 1] = hexArray[v & 0x0F];
        }
        return new String(hexChars);
    }
    
    private static byte[] hexToBytes(String s) {
        int len = s.length();
        if (len % 2 != 0) {
            throw new IllegalArgumentException("Illegal string length: "+len);
        }

        byte[] data = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            data[i / 2] = (byte) ((Character.digit(s.charAt(i), 16) << 4)
                                 + Character.digit(s.charAt(i+1), 16));
        }
        return data;
    }
    
    private X509Certificate extractPlainCert(String text) throws XmlPullParserException, IOException
    {
        // Plaincert may be null if another client is already trying to pair
        String certText = NvHTTP.getXmlString(text, "plaincert", false);
        if (certText != null) {
            byte[] certBytes = hexToBytes(certText);

            try {
                CertificateFactory cf = CertificateFactory.getInstance("X.509");
                return (X509Certificate)cf.generateCertificate(new ByteArrayInputStream(certBytes));
            } catch (CertificateException e) {
                e.printStackTrace();
                throw new RuntimeException(e);
            }
        }
        else {
            return null;
        }
    }
    
    private byte[] generateRandomBytes(int length)
    {
        byte[] rand = new byte[length];
        new SecureRandom().nextBytes(rand);
        return rand;
    }
    
    private static byte[] saltPin(byte[] salt, String pin) throws UnsupportedEncodingException {
        byte[] saltedPin = new byte[salt.length + pin.length()];
        System.arraycopy(salt, 0, saltedPin, 0, salt.length);
        System.arraycopy(pin.getBytes("UTF-8"), 0, saltedPin, salt.length, pin.length());
        return saltedPin;
    }

    private static Signature getSha256SignatureInstanceForKey(Key key) throws NoSuchAlgorithmException {
        switch (key.getAlgorithm()) {
            case "RSA":
                return Signature.getInstance("SHA256withRSA");
            case "EC":
                return Signature.getInstance("SHA256withECDSA");
            default:
                throw new NoSuchAlgorithmException("Unhandled key algorithm: " + key.getAlgorithm());
        }
    }
    
    private static boolean verifySignature(byte[] data, byte[] signature, Certificate cert) {
        try {
            Signature sig = PairingManager.getSha256SignatureInstanceForKey(cert.getPublicKey());
            sig.initVerify(cert.getPublicKey());
            sig.update(data);
            return sig.verify(signature);
        } catch (NoSuchAlgorithmException | SignatureException | InvalidKeyException e) {
            e.printStackTrace();
            throw new RuntimeException(e);
        }
    }
    
    private static byte[] signData(byte[] data, PrivateKey key) {
        try {
            Signature sig = PairingManager.getSha256SignatureInstanceForKey(key);
            sig.initSign(key);
            sig.update(data);
            return sig.sign();
        } catch (NoSuchAlgorithmException | SignatureException | InvalidKeyException e) {
            e.printStackTrace();
            throw new RuntimeException(e);
        }
    }

    private static byte[] performBlockCipher(BlockCipher blockCipher, byte[] input) {
        int blockSize = blockCipher.getBlockSize();
        int blockRoundedSize = (input.length + (blockSize - 1)) & ~(blockSize - 1);

        byte[] blockRoundedInputData = Arrays.copyOf(input, blockRoundedSize);
        byte[] blockRoundedOutputData = new byte[blockRoundedSize];

        for (int offset = 0; offset < blockRoundedSize; offset += blockSize) {
            blockCipher.processBlock(blockRoundedInputData, offset, blockRoundedOutputData, offset);
        }

        return blockRoundedOutputData;
    }
    
    private static byte[] decryptAes(byte[] encryptedData, byte[] aesKey) {
        BlockCipher aesEngine = new AESLightEngine();
        aesEngine.init(false, new KeyParameter(aesKey));
        return performBlockCipher(aesEngine, encryptedData);
    }
    
    private static byte[] encryptAes(byte[] plaintextData, byte[] aesKey) {
        BlockCipher aesEngine = new AESLightEngine();
        aesEngine.init(true, new KeyParameter(aesKey));
        return performBlockCipher(aesEngine, plaintextData);
    }
    
    private static byte[] generateAesKey(PairingHashAlgorithm hashAlgo, byte[] keyData) {
        return Arrays.copyOf(hashAlgo.hashData(keyData), 16);
    }
    
    private static byte[] concatBytes(byte[] a, byte[] b) {
        byte[] c = new byte[a.length + b.length];
        System.arraycopy(a, 0, c, 0, a.length);
        System.arraycopy(b, 0, c, a.length, b.length);
        return c;
    }
    
    public static String generatePinString() {
        SecureRandom r = new SecureRandom();
        return String.format((Locale)null, "%d%d%d%d",
                r.nextInt(10), r.nextInt(10),
                r.nextInt(10), r.nextInt(10));
    }

    public X509Certificate getPairedCert() {
        return serverCert;
    }
    
    /**
     * 配对处理
     * @param serverInfo 服务端信息
     * @param pin 配对PIN码
     * @return PairResult 包含配对状态和pairname
     * @throws IOException
     * @throws XmlPullParserException
     */
    public PairResult pair(String serverInfo, String pin) throws IOException, XmlPullParserException, InterruptedException {
        PairingHashAlgorithm hashAlgo;

        int serverMajorVersion = http.getServerMajorVersion(serverInfo);
        LimeLog.info("Pairing with server generation: " + serverMajorVersion);
        if (serverMajorVersion >= 7) {
            // Gen 7+ uses SHA-256 hashing
            hashAlgo = new Sha256PairingHash();
        } else {
            // Prior to Gen 7, SHA-1 is used
            hashAlgo = new Sha1PairingHash();
        }

        // 生成用于PIN哈希的salt
        byte[] salt = generateRandomBytes(16);

        // 合并salt和pin，然后用它们生成AES密钥
        byte[] aesKey = generateAesKey(hashAlgo, saltPin(salt, pin));

        // 发送salt并获取服务端证书。此处没有读取超时，因为用户必须输入PIN后服务端才会响应
        String getCert = http.executePairingCommand("phrase=getservercert&salt=" +
                bytesToHex(salt) + "&clientcert=" + bytesToHex(pemCertBytes),
                false);
        if (!NvHTTP.getXmlString(getCert, "paired", true).equals("1")) {
            return new PairResult(PairState.FAILED, null);
        }

        // 获取配对名（pairname），兼容服务端未返回pairname的情况
        String pairName = NvHTTP.getXmlString(getCert, "pairname", false);

        // 保存证书以便后续检索
        serverCert = extractPlainCert(getCert);
        if (serverCert == null) {
            // 如果有其他设备正在配对，GFE会返回空证书
            http.unpair();
            return new PairResult(PairState.ALREADY_IN_PROGRESS, pairName);
        }

        // 要求此证书用于与此主机的TLS通信
        http.setServerCert(serverCert);

        // 生成随机挑战并用AES密钥加密
        byte[] randomChallenge = generateRandomBytes(16);
        byte[] encryptedChallenge = encryptAes(randomChallenge, aesKey);

        // 发送加密挑战到服务端
        String challengeResp = http.executePairingCommand("clientchallenge=" + bytesToHex(encryptedChallenge), true);
        if (!NvHTTP.getXmlString(challengeResp, "paired", true).equals("1")) {
            http.unpair();
            return new PairResult(PairState.FAILED, pairName);
        }

        // 解码服务端响应和后续挑战
        byte[] encServerChallengeResponse = hexToBytes(NvHTTP.getXmlString(challengeResp, "challengeresponse", true));
        byte[] decServerChallengeResponse = decryptAes(encServerChallengeResponse, aesKey);

        byte[] serverResponse = Arrays.copyOfRange(decServerChallengeResponse, 0, hashAlgo.getHashLength());
        byte[] serverChallenge = Arrays.copyOfRange(decServerChallengeResponse, hashAlgo.getHashLength(), hashAlgo.getHashLength() + 16);

        // 用另一个16字节的secret，结合secret、证书签名和challenge计算挑战响应哈希
        byte[] clientSecret = generateRandomBytes(16);
        byte[] challengeRespHash = hashAlgo.hashData(concatBytes(concatBytes(serverChallenge, cert.getSignature()), clientSecret));
        byte[] challengeRespEncrypted = encryptAes(challengeRespHash, aesKey);
        String secretResp = http.executePairingCommand("serverchallengeresp=" + bytesToHex(challengeRespEncrypted), true);
        if (!NvHTTP.getXmlString(secretResp, "paired", true).equals("1")) {
            http.unpair();
            return new PairResult(PairState.FAILED, pairName);
        }

        // 获取服务端签名的secret
        byte[] serverSecretResp = hexToBytes(NvHTTP.getXmlString(secretResp, "pairingsecret", true));
        byte[] serverSecret = Arrays.copyOfRange(serverSecretResp, 0, 16);
        byte[] serverSignature = Arrays.copyOfRange(serverSecretResp, 16, serverSecretResp.length);

        // 校验数据真实性
        if (!verifySignature(serverSecret, serverSignature, serverCert)) {
            // 取消配对流程
            http.unpair();
            // 可能是中间人攻击
            return new PairResult(PairState.FAILED, pairName);
        }

        // 校验服务端challenge是否与预期一致（即PIN是否正确）
        byte[] serverChallengeRespHash = hashAlgo.hashData(concatBytes(concatBytes(randomChallenge, serverCert.getSignature()), serverSecret));
        if (!Arrays.equals(serverChallengeRespHash, serverResponse)) {
            // 取消配对流程
            http.unpair();
            // 可能PIN错误
            return new PairResult(PairState.PIN_WRONG, pairName);
        }

        // 发送客户端签名的secret给服务端
        byte[] clientPairingSecret = concatBytes(clientSecret, signData(clientSecret, pk));
        String clientSecretResp = http.executePairingCommand("clientpairingsecret=" + bytesToHex(clientPairingSecret), true);
        if (!NvHTTP.getXmlString(clientSecretResp, "paired", true).equals("1")) {
            http.unpair();
            return new PairResult(PairState.FAILED, pairName);
        }

        // 执行初始挑战（似乎有必要让我们显示为已配对）
        String pairChallenge = http.executePairingChallenge();
        if (!NvHTTP.getXmlString(pairChallenge, "paired", true).equals("1")) {
            http.unpair();
            return new PairResult(PairState.FAILED, pairName);
        }

        return new PairResult(PairState.PAIRED, pairName);
    }

    /**
     * 配对结果类，包含配对状态和pairname
     */
    public static class PairResult {
        public final PairState state;
        public final String pairName;

        public PairResult(PairState state, String pairName) {
            this.state = state;
            this.pairName = pairName != null && !pairName.equals("unknown") ? pairName : "";
        }
    }
    
    private interface PairingHashAlgorithm {
        int getHashLength();
        byte[] hashData(byte[] data);
    }
    
    private static class Sha1PairingHash implements PairingHashAlgorithm {
        public int getHashLength() {
            return 20;
        }
        
        public byte[] hashData(byte[] data) {
            try {
                MessageDigest md = MessageDigest.getInstance("SHA-1");
                return md.digest(data);
            }
            catch (NoSuchAlgorithmException e) {
                e.printStackTrace();
                throw new RuntimeException(e);
            }
        }
    }
    
    private static class Sha256PairingHash implements PairingHashAlgorithm {
        public int getHashLength() {
            return 32;
        }
        
        public byte[] hashData(byte[] data) {
            try {
                MessageDigest md = MessageDigest.getInstance("SHA-256");
                return md.digest(data);
            }
            catch (NoSuchAlgorithmException e) {
                e.printStackTrace();
                throw new RuntimeException(e);
            }
        }
    }
}

File: app/src/main/java/com/limelight/nvstream/http/PairingManager.kt
package com.limelight.nvstream.http

import org.bouncycastle.crypto.engines.AESLightEngine
import org.bouncycastle.crypto.params.KeyParameter

import org.xmlpull.v1.XmlPullParserException

import com.limelight.LimeLog

import java.io.*
import java.security.*
import java.security.cert.CertificateException
import java.security.cert.CertificateFactory
import java.security.cert.X509Certificate
import java.util.Arrays
import java.util.Locale

class PairingManager(
    private val http: NvHTTP,
    cryptoProvider: LimelightCryptoProvider
) {
    private val pk: PrivateKey = cryptoProvider.getClientPrivateKey()
    private val cert: X509Certificate = cryptoProvider.getClientCertificate()
    private val pemCertBytes: ByteArray = cryptoProvider.getPemEncodedClientCertificate()

    var pairedCert: X509Certificate? = null
        private set

    enum class PairState {
        NOT_PAIRED,
        PAIRED,
        PIN_WRONG,
        FAILED,
        ALREADY_IN_PROGRESS
    }

    class PairResult(val state: PairState, pairName: String?) {
        val pairName: String = if (pairName != null && pairName != "unknown") pairName else ""
    }

    /**
     * 配对处理
     */
    @Throws(IOException::class, XmlPullParserException::class, InterruptedException::class)
    fun pair(serverInfo: String, pin: String): PairResult {
        val hashAlgo: PairingHashAlgorithm

        val serverMajorVersion = http.getServerMajorVersion(serverInfo)
        LimeLog.info("Pairing with server generation: $serverMajorVersion")
        hashAlgo = if (serverMajorVersion >= 7) {
            Sha256PairingHash()
        } else {
            Sha1PairingHash()
        }

        val salt = generateRandomBytes(16)
        val aesKey = generateAesKey(hashAlgo, saltPin(salt, pin))

        val getCert = http.executePairingCommand(
            "phrase=getservercert&salt=${bytesToHex(salt)}&clientcert=${bytesToHex(pemCertBytes)}",
            false
        )

        val pairedValue: String? = try {
            NvHTTP.getXmlString(getCert, "paired", true)
        } catch (e: HostHttpResponseException) {
            LimeLog.warning("getservercert returned status ${e.getErrorCode()}, checking paired value anyway")
            extractXmlValue(getCert, "paired")
        }
        if (pairedValue == null || pairedValue != "1") {
            return PairResult(PairState.FAILED, null)
        }

        val pairName: String? = try {
            NvHTTP.getXmlString(getCert, "pairname", false)
        } catch (e: HostHttpResponseException) {
            extractXmlValue(getCert, "pairname")
        }

        val serverCert = extractPlainCert(getCert)
        if (serverCert == null) {
            http.unpair()
            return PairResult(PairState.ALREADY_IN_PROGRESS, pairName)
        }

        pairedCert = serverCert
        http.serverCert = serverCert

        val randomChallenge = generateRandomBytes(16)
        val encryptedChallenge = encryptAes(randomChallenge, aesKey)

        val challengeResp = http.executePairingCommand("clientchallenge=${bytesToHex(encryptedChallenge)}", true)
        if (NvHTTP.getXmlString(challengeResp, "paired", true) != "1") {
            http.unpair()
            return PairResult(PairState.FAILED, pairName)
        }

        val encServerChallengeResponse = hexToBytes(NvHTTP.getXmlString(challengeResp, "challengeresponse", true)!!)
        val decServerChallengeResponse = decryptAes(encServerChallengeResponse, aesKey)

        val serverResponse = decServerChallengeResponse.copyOfRange(0, hashAlgo.hashLength)
        val serverChallenge = decServerChallengeResponse.copyOfRange(hashAlgo.hashLength, hashAlgo.hashLength + 16)

        val clientSecret = generateRandomBytes(16)
        val challengeRespHash = hashAlgo.hashData(concatBytes(concatBytes(serverChallenge, cert.signature), clientSecret))
        val challengeRespEncrypted = encryptAes(challengeRespHash, aesKey)
        val secretResp = http.executePairingCommand("serverchallengeresp=${bytesToHex(challengeRespEncrypted)}", true)
        if (NvHTTP.getXmlString(secretResp, "paired", true) != "1") {
            http.unpair()
            return PairResult(PairState.FAILED, pairName)
        }

        val serverSecretResp = hexToBytes(NvHTTP.getXmlString(secretResp, "pairingsecret", true)!!)
        val serverSecret = serverSecretResp.copyOfRange(0, 16)
        val serverSignature = serverSecretResp.copyOfRange(16, serverSecretResp.size)

        if (!verifySignature(serverSecret, serverSignature, serverCert)) {
            http.unpair()
            return PairResult(PairState.FAILED, pairName)
        }

        val serverChallengeRespHash = hashAlgo.hashData(concatBytes(concatBytes(randomChallenge, serverCert.signature), serverSecret))
        if (!Arrays.equals(serverChallengeRespHash, serverResponse)) {
            http.unpair()
            return PairResult(PairState.PIN_WRONG, pairName)
        }

        val clientPairingSecret = concatBytes(clientSecret, signData(clientSecret, pk))
        val clientSecretResp = http.executePairingCommand("clientpairingsecret=${bytesToHex(clientPairingSecret)}", true)
        if (NvHTTP.getXmlString(clientSecretResp, "paired", true) != "1") {
            http.unpair()
            return PairResult(PairState.FAILED, pairName)
        }

        val pairChallenge = http.executePairingChallenge()
        if (NvHTTP.getXmlString(pairChallenge, "paired", true) != "1") {
            http.unpair()
            return PairResult(PairState.FAILED, pairName)
        }

        return PairResult(PairState.PAIRED, pairName)
    }

    @Throws(XmlPullParserException::class, IOException::class)
    private fun extractPlainCert(text: String): X509Certificate? {
        val certText: String? = try {
            NvHTTP.getXmlString(text, "plaincert", false)
        } catch (e: HostHttpResponseException) {
            extractXmlValue(text, "plaincert")
        }
        if (certText != null) {
            val certBytes = hexToBytes(certText)
            return try {
                val cf = CertificateFactory.getInstance("X.509")
                cf.generateCertificate(ByteArrayInputStream(certBytes)) as X509Certificate
            } catch (e: CertificateException) {
                e.printStackTrace()
                throw RuntimeException(e)
            }
        }
        return null
    }

    private fun generateRandomBytes(length: Int): ByteArray {
        val rand = ByteArray(length)
        SecureRandom().nextBytes(rand)
        return rand
    }

    private interface PairingHashAlgorithm {
        val hashLength: Int
        fun hashData(data: ByteArray): ByteArray
    }

    private class Sha1PairingHash : PairingHashAlgorithm {
        override val hashLength: Int = 20

        override fun hashData(data: ByteArray): ByteArray {
            return try {
                MessageDigest.getInstance("SHA-1").digest(data)
            } catch (e: NoSuchAlgorithmException) {
                e.printStackTrace()
                throw RuntimeException(e)
            }
        }
    }

    private class Sha256PairingHash : PairingHashAlgorithm {
        override val hashLength: Int = 32

        override fun hashData(data: ByteArray): ByteArray {
            return try {
                MessageDigest.getInstance("SHA-256").digest(data)
            } catch (e: NoSuchAlgorithmException) {
                e.printStackTrace()
                throw RuntimeException(e)
            }
        }
    }

    companion object {
        private val hexArray = "0123456789ABCDEF".toCharArray()

        private fun bytesToHex(bytes: ByteArray): String {
            val hexChars = CharArray(bytes.size * 2)
            for (j in bytes.indices) {
                val v = bytes[j].toInt() and 0xFF
                hexChars[j * 2] = hexArray[v ushr 4]
                hexChars[j * 2 + 1] = hexArray[v and 0x0F]
            }
            return String(hexChars)
        }

        private fun hexToBytes(s: String): ByteArray {
            val len = s.length
            require(len % 2 == 0) { "Illegal string length: $len" }

            val data = ByteArray(len / 2)
            for (i in 0 until len step 2) {
                data[i / 2] = ((Character.digit(s[i], 16) shl 4) + Character.digit(s[i + 1], 16)).toByte()
            }
            return data
        }

        @Throws(UnsupportedEncodingException::class)
        private fun saltPin(salt: ByteArray, pin: String): ByteArray {
            val pinBytes = pin.toByteArray(Charsets.UTF_8)
            val saltedPin = ByteArray(salt.size + pinBytes.size)
            System.arraycopy(salt, 0, saltedPin, 0, salt.size)
            System.arraycopy(pinBytes, 0, saltedPin, salt.size, pinBytes.size)
            return saltedPin
        }

        private fun getSha256SignatureInstanceForKey(key: Key): Signature {
            return when (key.algorithm) {
                "RSA" -> Signature.getInstance("SHA256withRSA")
                "EC" -> Signature.getInstance("SHA256withECDSA")
                else -> throw NoSuchAlgorithmException("Unhandled key algorithm: ${key.algorithm}")
            }
        }

        private fun verifySignature(data: ByteArray, signature: ByteArray, cert: java.security.cert.Certificate): Boolean {
            return try {
                val sig = getSha256SignatureInstanceForKey(cert.publicKey)
                sig.initVerify(cert.publicKey)
                sig.update(data)
                sig.verify(signature)
            } catch (e: Exception) {
                e.printStackTrace()
                throw RuntimeException(e)
            }
        }

        private fun signData(data: ByteArray, key: PrivateKey): ByteArray {
            return try {
                val sig = getSha256SignatureInstanceForKey(key)
                sig.initSign(key)
                sig.update(data)
                sig.sign()
            } catch (e: Exception) {
                e.printStackTrace()
                throw RuntimeException(e)
            }
        }

        private fun performBlockCipher(blockCipher: org.bouncycastle.crypto.BlockCipher, input: ByteArray): ByteArray {
            val blockSize = blockCipher.blockSize
            val blockRoundedSize = (input.size + (blockSize - 1)) and (blockSize - 1).inv()

            val blockRoundedInputData = Arrays.copyOf(input, blockRoundedSize)
            val blockRoundedOutputData = ByteArray(blockRoundedSize)

            var offset = 0
            while (offset < blockRoundedSize) {
                blockCipher.processBlock(blockRoundedInputData, offset, blockRoundedOutputData, offset)
                offset += blockSize
            }

            return blockRoundedOutputData
        }

        private fun decryptAes(encryptedData: ByteArray, aesKey: ByteArray): ByteArray {
            val aesEngine = AESLightEngine()
            aesEngine.init(false, KeyParameter(aesKey))
            return performBlockCipher(aesEngine, encryptedData)
        }

        private fun encryptAes(plaintextData: ByteArray, aesKey: ByteArray): ByteArray {
            val aesEngine = AESLightEngine()
            aesEngine.init(true, KeyParameter(aesKey))
            return performBlockCipher(aesEngine, plaintextData)
        }

        private fun generateAesKey(hashAlgo: PairingHashAlgorithm, keyData: ByteArray): ByteArray {
            return Arrays.copyOf(hashAlgo.hashData(keyData), 16)
        }

        private fun concatBytes(a: ByteArray, b: ByteArray): ByteArray {
            val c = ByteArray(a.size + b.size)
            System.arraycopy(a, 0, c, 0, a.size)
            System.arraycopy(b, 0, c, a.size, b.size)
            return c
        }

        fun generatePinString(): String {
            val r = SecureRandom()
            return String.format(null as Locale?, "%d%d%d%d",
                r.nextInt(10), r.nextInt(10),
                r.nextInt(10), r.nextInt(10))
        }

        private fun extractXmlValue(xml: String, tagName: String): String? {
            val startTag = "<$tagName>"
            val start = xml.indexOf(startTag)
            if (start < 0) return null
            val valueStart = start + startTag.length
            val end = xml.indexOf("</$tagName>", valueStart)
            if (end < 0) return null
            return xml.substring(valueStart, end)
        }
    }
}
